### DISCUSSION TITLE
Please add built-in upload protocols.

### DISCUSSION THREAD

**Discussion by XanCraft21** [stm32duino/Arduino_Core_STM32#2425] [Ideas]
Hello, i was wondering if you can add some features for convenience reasons. I know this has been talked about many times, but i’m in desperate need of these features.

Can you please add native built-in upload protocols into this arduino boards package, for SWD, Serial, and DFU modes?

i don’t want to use STM32CubeProgrammer to upload my code, i prefer using native interfacing for convenience and ease of access reasons. Before i had to modify the software to add native Serial upload, but i don’t really want to do that anymore.

if you agree to adding these solutions, i will be very happy. Thank you for taking the time for reading this, sorry for sounding desperate.

**Comment by fpistm** [comment]
Hi @XanCraft21 
Unfortunately, I can not add this for several reasons:
- dfu-utils is not compatible with all STM32 series
- same for tools to flash stm32 (ex: stm32flash,..) using serial,...
- Requires to build, packages for several host arch,... maintains and answer issues related to their usages and I don't have time for this.

To customize the core for your own purpose you can create your own core package support using the STM32 core as a dependency or add a boards.local.txt/platform.local.txt. Like this you can define your own tools.

Ref:
https://arduino.github.io/arduino-cli/0.35/platform-specification/#platformlocaltxt
https://arduino.github.io/arduino-cli/0.35/platform-specification/#boardslocaltxt
https://arduino.github.io/arduino-cli/0.35/platform-specification/#core-reference
https://arduino.github.io/arduino-cli/0.35/platform-specification/#tool-references

https://github.com/stm32duino/Arduino_Core_STM32/wiki/Custom-board-based-on-a-core

**Reply by XanCraft21** [reply]
Ok. Thank you for answering this.

Also i think this thread should be closed as not planned because you are unable to add the feature.
